recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,399 @@
+
+
+## Objectives
+
+**Reference for new type recipeConfig and handling of secrets:** [Design document for Private Terraform Repository](https://github.com/radius-project/design-notes/blob/3644b754152edc97e641f537b20cf3d87a386c43/recipe/2024-01-support-private-terraform-repository.md)
+
+> **Issue Reference:** <!-- (If appropriate) Reference an existing issue that describes the feature or bug. -->
+https://github.com/radius-project/radius/issues/6539
+
+### Goals
+
+1. Enable users to use terraform recipes with multiple providers (including and outside of Azure, Kubernetes, AWS).
+2. Enable users to use terraform recipes with multiple configurations for the same provider using *alias*. 
+3. Enable provider configurations to be available for all recipies used in an environment.
+
+
+### Non goals
+1. Updates to Bicep Provider configuration.
+2. Authentication for providers hosted in private registries.
+3. Support for .terraformrc files.
+
+### User scenarios (optional)
+
+#### User story 1
+As an operator, I maintain a set of Terraform recipes for use with Radius. I have a set of provider configurations which are applicable across multiple recipes and I would like to configure them in a single centralized location for ease of maintenance.
+
+
+#### User story 2
+As an operator, I would like to manage cloud resources in different geographical regions. To enable this, I need to set up multiple configurations for the same provider for different regions and use an alias to refer to individual provider configurations. 
+
+## Design
+### Design details
+
+Reviewing popular provider configurations (GCP, Oracle, Heroku, DigitalOcean, Docker etc.) a large number of provider configurations can be setup by handling a combination of key value pairs, key value pairs in nested objects, secrets and environment variables.
+
+### Key-value pairs
+Key-value pairs will be parsed and saved to a Terraform configuration file.
+
+### Secrets
+Secrets will be handled similarly to the approach described in document [Design document for Private Terraform Repository](https://github.com/radius-project/design-notes/blob/3644b754152edc97e641f537b20cf3d87a386c43/recipe/2024-01-support-private-terraform-repository.md) wherein Applications.Core/secretStores can point to an existing K8s secret.
+
+The system will call the ListSecrets() api in Applications.Core namespace, retrieve contents of the secret and build the Terraform provider configuration.
+
+### Environment variables
+In a significant number of providers, as per documentation, environment variables are used one of the methods of saving sensitive credential data along with insensitive data. We allow the users to set environment variables for provider configuration. For sensitive information, we recommend the users save these values as secrets and point to them inside the env block.
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+     providers: [...]
+     env: {
+      key: value
+      envSecrets: [
```

Comment:
If we're adding a `valueFrom` property below, where the user can specify a secret store, does the user also have to specify that this is a secret under `envSecrets`? Put another way, what if `valueFrom` was directly underneath `env`?

Also, what is the `key` on line 69? Is that the key for the environment variable? How does that relate to the environment variable key that appears to be on lines 72 and 80?

Reply:
Line 187 is more helpful to understand. I need to update the description. will do soon.
Current proposal is to provide 3 ways of setting environment variables:
1. the key would be name of env var and value is the value set (non sensitive values)
2. envSecrets : Individual Secrets from SecretStore
3. envSecretsFrom: All Secrets from Secret Store
All three are optional.

Reply:
schema is updated per mtg discussion